File: src/features/auth/data/sessionRepository.ts
import type { KeyValueStore } from '../../../core/storage/keyValueStore';

export interface Session { readonly userId: string; readonly phone: string }

export interface SessionRepository {
  restore(): Session | null;
  save(userId: string, phone: string): Promise<void>;
  clear(): Promise<void>;
}

const USER_ID_KEY = 'session.userId';
const PHONE_KEY = 'session.phone';

export class StoredSessionRepository implements SessionRepository {
  constructor(private readonly store: KeyValueStore) {}

  restore(): Session | null {
    const userId = this.store.read(USER_ID_KEY);
    const phone = this.store.read(PHONE_KEY);
    if (userId === null || phone === null) return null;
    return { userId, phone };
  }

  async save(userId: string, phone: string): Promise<void> {
    await this.store.write(USER_ID_KEY, userId);
    await this.store.write(PHONE_KEY, phone);
  }

  async clear(): Promise<void> {
    await this.store.remove(USER_ID_KEY);
    await this.store.remove(PHONE_KEY);
  }
}

File: lib/features/auth/data/session_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/storage/key_value_store.dart';

abstract interface class SessionRepository {
  ({String userId, String phone})? restore();
  Future<void> save(String userId, String phone);
  Future<void> clear();
}

class StoredSessionRepository implements SessionRepository {
  StoredSessionRepository(this._store);

  final KeyValueStore _store;

  static const _userIdKey = 'session.userId';
  static const _phoneKey = 'session.phone';

  @override
  ({String userId, String phone})? restore() {
    final userId = _store.read(_userIdKey);
    final phone = _store.read(_phoneKey);
    if (userId == null || phone == null) return null;
    return (userId: userId, phone: phone);
  }

  @override
  Future<void> save(String userId, String phone) async {
    await _store.write(_userIdKey, userId);
    await _store.write(_phoneKey, phone);
  }

  @override
  Future<void> clear() async {
    await _store.remove(_userIdKey);
    await _store.remove(_phoneKey);
  }
}

final sessionRepositoryProvider = Provider<SessionRepository>(
  (ref) => StoredSessionRepository(ref.watch(keyValueStoreProvider)),
);
